### DISCUSSION TITLE
issue: Watched field state is different to rendered state with a custom resolver

### DISCUSSION THREAD

**Discussion by markjaniczak** [react-hook-form/react-hook-form#9977] [Q&A]
### Version Number

7.43.1

### Codesandbox/Expo snack

https://codesandbox.io/s/vigilant-tdd-1l9y13?file=/src/App.tsx

### Steps to reproduce

1. Enter a value for the _name_ field
2. Change the value of the _value_ field to below 5
3. Change the value of the _value_ field to 5 or above

### Expected behaviour

After step one you should see the message "Section of form is complete", after step two the message should be removed and after step 3 the message should appear again. The watch effect is checking if both the fields are valid and changing the state.

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Edge

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
I am not quite following this issue, could you provide some more details? are you saving the watch callback returned form values is different from the resolver?

**Comment by markjaniczak** [comment]
The watch callback checks that all fields are valid and changes the state to true / false depending on the outcome. In that example I am displaying the error message inline on each field using the field's state from the Controller component.

In the example I would expect to only see the completed message if all fields are valid but if you follow the steps you'll reach a point where one or both fields are invalid but the message appears which contradict each other:
![Screenshot 2023-02-20 at 2 38 49 pm](https://user-images.githubusercontent.com/13617683/220010822-c44739cb-e531-454c-a56f-5b2cb80b8126.png)

So I guess I'm asking; is it valid to call getFieldState within the watch callback? Because it seems to return the previous field state even when I add formState and getFieldState to the effect's dependencies.

**Comment by bluebill1049** [comment]
Thanks for added context and information, I think this maybe difficult as form values update first then followed by another relevant state update eg `isValid`, especially since validation could be async by nature. So `watch` may need to separate with form state update such as `isValid`.

**Comment by leapful** [comment]
@markjaniczak 

The codesandbox implementation has been changed. It would be great if you can provide the previous one with your issue.

Anyway, since you are using custom resolver, it will be easier to check all fields are valid inside custom resolver because at that time, data to validated is the whole form values.
